plugin/src/main/java/org/opensearch/ml/action/mcpserver/McpToolsHelper.java

```diff
@@ -184,24 +141,59 @@ public void searchAllToolsWithVersion(ActionListener<Map<String, Tuple<McpToolRe
                 log.error(errMsg, e);
                 restoreListener.onFailure(new OpenSearchException(errMsg));
             });
+
             client.search(buildSearchRequest(), actionListener);
         } catch (Exception e) {
             log.error("Failed to search mcp tools index", e);
             listener.onFailure(e);
         }
     }
 
-    public void searchAllTools(ActionListener<List<McpToolRegisterInput>> listener) {
+    public void searchToolsWithVersion(List<String> toolNames, ActionListener<List<McpToolRegisterInput>> listener) {
+        ActionListener<SearchResponse> actionListener = createSearchResponseListener(listener);
+        SearchRequest searchRequest = buildSearchRequest(toolNames);
+        searchRequest.source().version(true);
+        client.search(searchRequest, actionListener);
+    }
+
+    private ActionListener<SearchResponse> createSearchResponseListener(ActionListener<List<McpToolRegisterInput>> listener) {
+        return ActionListener.wrap(r -> {
+            List<McpToolRegisterInput> mcpTools = new ArrayList<>();
+            Arrays.stream(Objects.requireNonNull(r.getHits().getHits())).forEach(x -> {
+                try {
+                    McpToolRegisterInput mcpTool = parseMcpTool(x.getSourceAsString());
+                    mcpTools.add(mcpTool);
+                } catch (IOException e) {
+                    listener.onFailure(e);
```

Comment:
If have multiple MCP tools parsing failed, will call this `listener.onFailure(e);` multiple times. Is that ok ? I remember that will trigger some exception